funnel/models/event.py

```diff
@@ -0,0 +1,231 @@
+    db.Column('created_at', db.DateTime, default=datetime.utcnow, nullable=False)
+    )
+
+
+class Event(BaseScopedNameMixin, db.Model):
+    """
+    A discrete event under a proposal space.
+    For instance, a space could be associated with a workshop and a two-day conference.
+    The workshop constitutes as one event and each day of the conference
+    constitutes as an independent event.
+    This is done to allow distinguishing participants based on
+    on the tickets they have, given a participant may have a ticket
+    for only the workshop or a single day of the conference.
+    An event is associated with multiple ticket types,
+    which helps make the distinction between participants.
+    """
+    __tablename__ = 'event'
+
+    proposal_space_id = db.Column(None, db.ForeignKey('proposal_space.id'), nullable=False)
+    proposal_space = db.relationship(ProposalSpace,
+        backref=db.backref('events', cascade='all, delete-orphan', lazy='dynamic'))
+    parent = db.synonym('proposal_space')
+    ticket_types = db.relationship("TicketType", secondary=event_ticket_type, lazy='dynamic')
+    participants = db.relationship("Participant", secondary='attendee', backref="events")
+    __table_args__ = (db.UniqueConstraint('proposal_space_id', 'name'),)
+
+    @classmethod
+    def get(cls, space, title):
+        return cls.query.filter_by(title=title, proposal_space=space).first()
+
+    @classmethod
+    def create_from(cls, space, title):
+        event = cls(title=title, proposal_space=space)
+        db.session.add(event)
+        return event
+
+    @classmethod
+    def get_or_create_from(cls, space, title):
+        event = cls.get(space, title)
+        if not event:
+            event = cls.create_from(space, title)
+        return event
+
+
+class TicketType(BaseScopedNameMixin, db.Model):
+    """
+    Models different types of tickets. Eg: Early Geek, Super Early Geek, Workshop A.
+    A ticket type is associated with multiple events.
+    """
+    __tablename__ = 'ticket_type'
+
+    proposal_space_id = db.Column(None, db.ForeignKey('proposal_space.id'), nullable=False)
+    proposal_space = db.relationship(ProposalSpace,
+        backref=db.backref('ticket_types', cascade='all, delete-orphan', lazy='dynamic'))
+    parent = db.synonym('proposal_space')
+    events = db.relationship("Event", secondary=event_ticket_type)
+    __table_args__ = (db.UniqueConstraint('proposal_space_id', 'name'),)
+
+    @classmethod
+    def get(cls, space, title):
+        return cls.query.filter_by(title=title, proposal_space=space).first()
+
+    @classmethod
+    def create_from(cls, space, title, events=[]):
+        ticket_type = cls(
+            title=title,
+            proposal_space=space,
+            events=events
+        )
+        db.session.add(ticket_type)
+        return ticket_type
+
+    @classmethod
+    def get_or_create_from(cls, space, title, events=[]):
```

Comment:
If `events` is ignored when the record already exists, that is a problem. Your method name and usage pattern should be "create or update and return it" (in SQL, an UPSERT).